File: sdks/rust_sdk/src/audit.rs
// Copyright 2025 Rahul Rajaram
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Audit proof extension for SW4RM protocol.
//!
//! This module provides audit trail capabilities with optional proof generation
//! and verification, inspired by ZK-MCP research.

use sha2::{Digest, Sha256};
use std::collections::HashMap;
use uuid::Uuid;

use crate::envelope::EnvelopeData;

/// Represents a cryptographic or logical proof for an audit event.
#[derive(Debug, Clone, PartialEq)]
pub struct AuditProof {
    /// Unique identifier for this proof
    pub proof_id: String,
    /// Type of proof (e.g., "simple_hash", "zk_proof", "signature", "noop")
    pub proof_type: String,
    /// The actual proof data as bytes
    pub proof_data: Vec<u8>,
    /// Timestamp when the proof was created (HLC or unix ms)
    pub created_at: String,
    /// Whether this proof has been verified
    pub verified: bool,
}

impl AuditProof {
    /// Create a new AuditProof
    pub fn new(
        proof_id: String,
        proof_type: String,
        proof_data: Vec<u8>,
        created_at: String,
    ) -> Self {
        Self {
            proof_id,
            proof_type,
            proof_data,
            created_at,
            verified: false,
        }
    }

    /// Create a no-op proof (always verified)
    pub fn noop() -> Self {
        Self {
            proof_id: Uuid::new_v4().to_string(),
            proof_type: "noop".to_string(),
            proof_data: Vec::new(),
            created_at: chrono::Utc::now().timestamp_millis().to_string(),
            verified: true,
        }
    }
}

/// Defines audit requirements for envelope processing.
#[derive(Debug, Clone, PartialEq)]
pub struct AuditPolicy {
    /// Unique identifier for this policy
    pub policy_id: String,
    /// Whether proof is required for this policy
    pub require_proof: bool,
    /// Level of verification required (e.g., "none", "basic", "strict")
    pub verification_level: String,
    /// How many days to retain audit records
    pub retention_days: i32,
}

impl Default for AuditPolicy {
    fn default() -> Self {
        Self {
            policy_id: String::new(),
            require_proof: false,
            verification_level: "none".to_string(),
            retention_days: 90,
        }
    }
}

/// Represents a complete audit record for an envelope action.
#[derive(Debug, Clone, PartialEq)]
pub struct AuditRecord {
    /// Unique identifier for this audit record
    pub record_id: String,
    /// The message_id of the envelope being audited
    pub envelope_id: String,
    /// The action performed (e.g., "send", "receive", "process")
    pub action: String,
    /// ID of the agent/component that performed the action
    pub actor_id: String,
    /// When the action occurred (HLC or unix ms)
    pub timestamp: String,
    /// Optional proof associated with this action
    pub proof: Option<AuditProof>,
}

impl AuditRecord {
    /// Create a new AuditRecord
    pub fn new(
        envelope_id: String,
        action: String,
        actor_id: String,
        proof: Option<AuditProof>,
    ) -> Self {
        Self {
            record_id: Uuid::new_v4().to_string(),
            envelope_id,
            action,
            actor_id,
            timestamp: chrono::Utc::now().timestamp_millis().to_string(),
            proof,
        }
    }
}

/// Trait for audit implementations
pub trait Auditor: Send + Sync {
    /// Create a proof for an envelope action
    fn create_proof(&self, envelope: &EnvelopeData, action: &str) -> AuditProof;

    /// Verify a proof
    fn verify_proof(&self, proof: &AuditProof) -> bool;

    /// Record an audit event
    fn record(
        &mut self,
        envelope: &EnvelopeData,
        action: &str,
        proof: Option<AuditProof>,
    ) -> AuditRecord;

    /// Query audit records for an envelope
    fn query(&self, envelope_id: &str) -> Vec<AuditRecord>;
}

/// No-op auditor implementation (always succeeds, doesn't store)
#[derive(Debug, Default)]
pub struct NoOpAuditor;

impl NoOpAuditor {
    pub fn new() -> Self {
        Self
    }
}

impl Auditor for NoOpAuditor {
    fn create_proof(&self, _envelope: &EnvelopeData, _action: &str) -> AuditProof {
        AuditProof::noop()
    }

    fn verify_proof(&self, _proof: &AuditProof) -> bool {
        true
    }

    fn record(
        &mut self,
        envelope: &EnvelopeData,
        action: &str,
        proof: Option<AuditProof>,
    ) -> AuditRecord {
        AuditRecord::new(
            envelope.message_id.clone(),
            action.to_string(),
            envelope.producer_id.clone(),
            proof,
        )
    }

    fn query(&self, _envelope_id: &str) -> Vec<AuditRecord> {
        Vec::new()
    }
}

/// In-memory auditor implementation with simple hash proofs
#[derive(Debug, Default)]
pub struct InMemoryAuditor {
    records: HashMap<String, Vec<AuditRecord>>,
}

impl InMemoryAuditor {
    pub fn new() -> Self {
        Self {
            records: HashMap::new(),
        }
    }

    /// Get all records across all envelopes
    pub fn get_all_records(&self) -> Vec<AuditRecord> {
        self.records
            .values()
            .flatten()
            .cloned()
            .collect()
    }

    /// Clear all audit records
    pub fn clear(&mut self) {
        self.records.clear();
    }
}

impl Auditor for InMemoryAuditor {
    fn create_proof(&self, envelope: &EnvelopeData, _action: &str) -> AuditProof {
        create_simple_proof(envelope)
    }

    fn verify_proof(&self, proof: &AuditProof) -> bool {
        verify_audit_proof(proof)
    }

    fn record(
        &mut self,
        envelope: &EnvelopeData,
        action: &str,
        proof: Option<AuditProof>,
    ) -> AuditRecord {
        let record = AuditRecord::new(
            envelope.message_id.clone(),
            action.to_string(),
            envelope.producer_id.clone(),
            proof,
        );

        self.records
            .entry(envelope.message_id.clone())
            .or_default()
            .push(record.clone());

        record
    }

    fn query(&self, envelope_id: &str) -> Vec<AuditRecord> {
        self.records
            .get(envelope_id)
            .cloned()
            .unwrap_or_default()
    }
}

/// Compute a deterministic hash of an envelope
pub fn compute_envelope_hash(envelope: &EnvelopeData) -> String {
    let mut hasher = Sha256::new();
    hasher.update(envelope.message_id.as_bytes());
    hasher.update(envelope.producer_id.as_bytes());
    hasher.update(&envelope.payload);
    hasher.update(envelope.message_type.to_le_bytes());
    let result = hasher.finalize();
    hex::encode(result)
}

/// Create a simple hash-based proof for an envelope
pub fn create_simple_proof(envelope: &EnvelopeData) -> AuditProof {
    let hash = compute_envelope_hash(envelope);
    let proof_data = hex::decode(&hash).unwrap_or_default();

    AuditProof {
        proof_id: Uuid::new_v4().to_string(),
        proof_type: "simple_hash".to_string(),
        proof_data,
        created_at: chrono::Utc::now().timestamp_millis().to_string(),
        verified: false,
    }
}

/// Verify an audit proof
pub fn verify_audit_proof(proof: &AuditProof) -> bool {
    match proof.proof_type.as_str() {
        "noop" => true,
        "simple_hash" => proof.proof_data.len() == 32, // SHA256 output is 32 bytes
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::constants::message_type;
    use crate::envelope::EnvelopeBuilder;

    /// Helper function to create a test envelope
    fn create_test_envelope(producer_id: &str) -> EnvelopeData {
        EnvelopeBuilder::new(producer_id.to_string(), message_type::DATA).build()
    }

    /// Helper function to create a test envelope with payload
    fn create_test_envelope_with_payload(producer_id: &str, payload: Vec<u8>) -> EnvelopeData {
        EnvelopeBuilder::new(producer_id.to_string(), message_type::DATA)
            .with_payload(payload)
            .build()
    }

    /// Test AuditProof creation
    #[test]
    fn test_audit_proof_creation() {
        let proof = AuditProof::new(
            "proof-123".to_string(),
            "simple_hash".to_string(),
            b"test_proof_data".to_vec(),
            "1234567890".to_string(),
        );

        assert_eq!(proof.proof_id, "proof-123");
        assert_eq!(proof.proof_type, "simple_hash");
        assert_eq!(proof.proof_data, b"test_proof_data");
        assert_eq!(proof.created_at, "1234567890");
        assert!(!proof.verified);
    }

    /// Test AuditProof noop creation
    #[test]
    fn test_audit_proof_noop() {
        let proof = AuditProof::noop();

        assert_eq!(proof.proof_type, "noop");
        assert!(proof.proof_data.is_empty());
        assert!(proof.verified);
        assert!(!proof.proof_id.is_empty());
    }

    /// Test AuditPolicy creation
    #[test]
    fn test_audit_policy_creation() {
        let policy = AuditPolicy {
            policy_id: "policy-456".to_string(),
            require_proof: true,
            verification_level: "strict".to_string(),
            retention_days: 90,
        };

        assert_eq!(policy.policy_id, "policy-456");
        assert!(policy.require_proof);
        assert_eq!(policy.verification_level, "strict");
        assert_eq!(policy.retention_days, 90);
    }

    /// Test AuditPolicy default values
    #[test]
    fn test_audit_policy_default() {
        let policy = AuditPolicy::default();

        assert!(policy.policy_id.is_empty());
        assert!(!policy.require_proof);
        assert_eq!(policy.verification_level, "none");
        assert_eq!(policy.retention_days, 90);
    }

    /// Test AuditRecord creation
    #[test]
    fn test_audit_record_creation() {
        let proof = AuditProof::noop();
        let record = AuditRecord::new(
            "env-123".to_string(),
            "send".to_string(),
            "agent-1".to_string(),
            Some(proof.clone()),
        );

        assert!(!record.record_id.is_empty());
        assert_eq!(record.envelope_id, "env-123");
        assert_eq!(record.action, "send");
        assert_eq!(record.actor_id, "agent-1");
        assert!(!record.timestamp.is_empty());
        assert!(record.proof.is_some());
    }

    /// Test AuditRecord without proof
    #[test]
    fn test_audit_record_without_proof() {
        let record = AuditRecord::new(
            "env-123".to_string(),
            "send".to_string(),
            "agent-1".to_string(),
            None,
        );

        assert!(record.proof.is_none());
    }

    /// Test compute_envelope_hash is deterministic
    #[test]
    fn test_compute_envelope_hash_deterministic() {
        let envelope = create_test_envelope_with_payload("agent-1", b"test payload".to_vec());

        let hash1 = compute_envelope_hash(&envelope);
        let hash2 = compute_envelope_hash(&envelope);

        assert_eq!(hash1, hash2);
        assert_eq!(hash1.len(), 64); // SHA256 hex digest
    }

    /// Test compute_envelope_hash different for different envelopes
    #[test]
    fn test_compute_envelope_hash_different() {
        let envelope1 = create_test_envelope_with_payload("agent-1", b"payload1".to_vec());
        let envelope2 = create_test_envelope_with_payload("agent-2", b"payload2".to_vec());

        let hash1 = compute_envelope_hash(&envelope1);
        let hash2 = compute_envelope_hash(&envelope2);

        assert_ne!(hash1, hash2);
    }

    /// Test create_simple_proof
    #[test]
    fn test_create_simple_proof() {
        let envelope = create_test_envelope_with_payload("agent-1", b"test".to_vec());

        let proof = create_simple_proof(&envelope);

        assert_eq!(proof.proof_type, "simple_hash");
        assert_eq!(proof.proof_data.len(), 32); // SHA256 output
        assert!(!proof.verified);
        assert!(!proof.proof_id.is_empty());
        assert!(!proof.created_at.is_empty());
    }

    /// Test verify_audit_proof simple_hash
    #[test]
    fn test_verify_audit_proof_simple_hash() {
        let envelope = create_test_envelope_with_payload("agent-1", b"test".to_vec());

        let proof = create_simple_proof(&envelope);

        assert!(verify_audit_proof(&proof));
    }

    /// Test verify_audit_proof noop
    #[test]
    fn test_verify_audit_proof_noop() {
        let proof = AuditProof::noop();
        assert!(verify_audit_proof(&proof));
    }

    /// Test verify_audit_proof invalid simple_hash
    #[test]
    fn test_verify_audit_proof_invalid_simple_hash() {
        let proof = AuditProof::new(
            "test-id".to_string(),
            "simple_hash".to_string(),
            b"invalid".to_vec(), // Wrong size
            "1234567890".to_string(),
        );

        assert!(!verify_audit_proof(&proof));
    }

    /// Test verify_audit_proof unknown type
    #[test]
    fn test_verify_audit_proof_unknown_type() {
        let proof = AuditProof::new(
            "test-id".to_string(),
            "unknown_type".to_string(),
            b"some data".to_vec(),
            "1234567890".to_string(),
        );

        assert!(!verify_audit_proof(&proof));
    }

    /// Test NoOpAuditor create_proof
    #[test]
    fn test_noop_auditor_create_proof() {
        let auditor = NoOpAuditor::new();
        let envelope = create_test_envelope("agent-1");

        let proof = auditor.create_proof(&envelope, "send");

        assert_eq!(proof.proof_type, "noop");
        assert!(proof.proof_data.is_empty());
        assert!(proof.verified);
    }

    /// Test NoOpAuditor verify_proof always true
    #[test]
    fn test_noop_auditor_verify_proof() {
        let auditor = NoOpAuditor::new();
        let proof = AuditProof::new(
            "test".to_string(),
            "anything".to_string(),
            b"anything".to_vec(),
            "123".to_string(),
        );

        assert!(auditor.verify_proof(&proof));
    }

    /// Test NoOpAuditor record creates record
    #[test]
    fn test_noop_auditor_record() {
        let mut auditor = NoOpAuditor::new();
        let envelope = create_test_envelope("agent-1");

        let record = auditor.record(&envelope, "send", None);

        assert_eq!(record.envelope_id, envelope.message_id);
        assert_eq!(record.action, "send");
        assert_eq!(record.actor_id, "agent-1");
        assert!(record.proof.is_none());
    }

    /// Test NoOpAuditor query always empty
    #[test]
    fn test_noop_auditor_query() {
        let auditor = NoOpAuditor::new();
        assert!(auditor.query("any-id").is_empty());
    }

    /// Test InMemoryAuditor create_proof
    #[test]
    fn test_inmemory_auditor_create_proof() {
        let auditor = InMemoryAuditor::new();
        let envelope = create_test_envelope("agent-1");

        let proof = auditor.create_proof(&envelope, "send");

        assert_eq!(proof.proof_type, "simple_hash");
        assert_eq!(proof.proof_data.len(), 32);
        assert!(!proof.verified);
    }

    /// Test InMemoryAuditor verify_proof simple_hash
    #[test]
    fn test_inmemory_auditor_verify_proof() {
        let auditor = InMemoryAuditor::new();
        let envelope = create_test_envelope("agent-1");

        let proof = auditor.create_proof(&envelope, "send");
        assert!(auditor.verify_proof(&proof));
    }

    /// Test InMemoryAuditor verify_proof noop
    #[test]
    fn test_inmemory_auditor_verify_noop() {
        let auditor = InMemoryAuditor::new();
        let proof = AuditProof::noop();
        assert!(auditor.verify_proof(&proof));
    }

    /// Test InMemoryAuditor verify_proof invalid
    #[test]
    fn test_inmemory_auditor_verify_invalid() {
        let auditor = InMemoryAuditor::new();
        let proof = AuditProof::new(
            "test".to_string(),
            "unknown".to_string(),
            b"data".to_vec(),
            "123".to_string(),
        );

        assert!(!auditor.verify_proof(&proof));
    }

    /// Test InMemoryAuditor record stores record
    #[test]
    fn test_inmemory_auditor_record() {
        let mut auditor = InMemoryAuditor::new();
        let envelope = create_test_envelope("agent-1");

        let record = auditor.record(&envelope, "send", None);

        assert_eq!(record.envelope_id, envelope.message_id);
        assert_eq!(record.action, "send");

        // Query should return the record
        let records = auditor.query(&envelope.message_id);
        assert_eq!(records.len(), 1);
        assert_eq!(records[0], record);
    }

    /// Test InMemoryAuditor record multiple actions
    #[test]
    fn test_inmemory_auditor_multiple_actions() {
        let mut auditor = InMemoryAuditor::new();
        let envelope = create_test_envelope("agent-1");

        auditor.record(&envelope, "send", None);
        auditor.record(&envelope, "receive", None);
        auditor.record(&envelope, "process", None);

        let records = auditor.query(&envelope.message_id);
        assert_eq!(records.len(), 3);
        assert_eq!(records[0].action, "send");
        assert_eq!(records[1].action, "receive");
        assert_eq!(records[2].action, "process");
    }

    /// Test InMemoryAuditor query nonexistent envelope
    #[test]
    fn test_inmemory_auditor_query_nonexistent() {
        let auditor = InMemoryAuditor::new();
        let records = auditor.query("nonexistent-id");
        assert!(records.is_empty());
    }

    /// Test InMemoryAuditor get_all_records
    #[test]
    fn test_inmemory_auditor_get_all_records() {
        let mut auditor = InMemoryAuditor::new();
        let envelope1 = create_test_envelope("agent-1");
        let envelope2 = create_test_envelope("agent-2");

        auditor.record(&envelope1, "send", None);
        auditor.record(&envelope1, "receive", None);
        auditor.record(&envelope2, "send", None);

        let all_records = auditor.get_all_records();
        assert_eq!(all_records.len(), 3);
    }

    /// Test InMemoryAuditor clear
    #[test]
    fn test_inmemory_auditor_clear() {
        let mut auditor = InMemoryAuditor::new();
        let envelope = create_test_envelope("agent-1");

        auditor.record(&envelope, "send", None);
        assert_eq!(auditor.get_all_records().len(), 1);

        auditor.clear();
        assert!(auditor.get_all_records().is_empty());
        assert!(auditor.query(&envelope.message_id).is_empty());
    }

    /// Test full audit workflow
    #[test]
    fn test_full_audit_workflow() {
        let mut auditor = InMemoryAuditor::new();

        // Build envelope
        let envelope = create_test_envelope_with_payload("agent-1", b"test payload".to_vec());

        // Create proof
        let proof = auditor.create_proof(&envelope, "send");
        assert!(auditor.verify_proof(&proof));

        // Record action with proof
        let record = auditor.record(&envelope, "send", Some(proof.clone()));
        assert!(record.proof.is_some());

        // Query records
        let records = auditor.query(&envelope.message_id);
        assert_eq!(records.len(), 1);
        assert!(records[0].proof.is_some());
    }

    /// Test audit trail across multiple agents
    #[test]
    fn test_audit_trail_multiple_agents() {
        let mut auditor = InMemoryAuditor::new();

        // Agent 1 sends
        let envelope = create_test_envelope("agent-1");

        let proof1 = auditor.create_proof(&envelope, "send");
        auditor.record(&envelope, "send", Some(proof1));

        // Agent 2 receives
        let proof2 = auditor.create_proof(&envelope, "receive");
        auditor.record(&envelope, "receive", Some(proof2));

        // Agent 2 processes
        let proof3 = auditor.create_proof(&envelope, "process");
        auditor.record(&envelope, "process", Some(proof3));

        // Verify audit trail
        let records = auditor.query(&envelope.message_id);
        assert_eq!(records.len(), 3);
        assert_eq!(records[0].action, "send");
        assert_eq!(records[1].action, "receive");
        assert_eq!(records[2].action, "process");

        // Each has a proof
        for record in &records {
            assert!(record.proof.is_some());
            assert!(auditor.verify_proof(record.proof.as_ref().unwrap()));
        }
    }

    /// Test AuditProof clone
    #[test]
    fn test_audit_proof_clone() {
        let proof = AuditProof::new(
            "proof-1".to_string(),
            "simple_hash".to_string(),
            b"data".to_vec(),
            "123".to_string(),
        );

        let cloned = proof.clone();
        assert_eq!(proof, cloned);
    }

    /// Test AuditPolicy clone
    #[test]
    fn test_audit_policy_clone() {
        let policy = AuditPolicy {
            policy_id: "policy-1".to_string(),
            require_proof: true,
            verification_level: "strict".to_string(),
            retention_days: 30,
        };

        let cloned = policy.clone();
        assert_eq!(policy, cloned);
    }

    /// Test AuditRecord clone
    #[test]
    fn test_audit_record_clone() {
        let record = AuditRecord::new(
            "env-1".to_string(),
            "send".to_string(),
            "agent-1".to_string(),
            None,
        );

        let cloned = record.clone();
        assert_eq!(record.envelope_id, cloned.envelope_id);
        assert_eq!(record.action, cloned.action);
        assert_eq!(record.actor_id, cloned.actor_id);
    }

    /// Test proof with different verification levels
    #[test]
    fn test_verification_levels() {
        let levels = vec!["none", "basic", "strict"];

        for level in levels {
            let policy = AuditPolicy {
                policy_id: format!("policy-{}", level),
                require_proof: level != "none",
                verification_level: level.to_string(),
                retention_days: 90,
            };

            assert_eq!(policy.verification_level, level);
        }
    }

    /// Test different retention periods
    #[test]
    fn test_retention_periods() {
        let periods = vec![7, 30, 90, 365];

        for days in periods {
            let policy = AuditPolicy {
                policy_id: "policy-1".to_string(),
                require_proof: false,
                verification_level: "none".to_string(),
                retention_days: days,
            };

            assert_eq!(policy.retention_days, days);
        }
    }
}

File: sdks/js_sdk/src/audit.ts
// Copyright 2025 Rahul Rajaram
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Audit proof extension for SW4RM protocol.
 *
 * This module provides audit trail capabilities with optional proof generation
 * and verification, inspired by ZK-MCP research.
 */

import crypto from 'node:crypto';
import { nowHlcStub } from './internal/envelope.js';
import { EnvelopeState } from './constants/index.js';

/**
 * Represents a cryptographic or logical proof for an audit event.
 */
export interface AuditProof {
  /** Unique identifier for this proof */
  proof_id: string;
  /** Type of proof (e.g., "simple_hash", "zk_proof", "signature") */
  proof_type: string;
  /** The actual proof data as bytes */
  proof_data: Uint8Array;
  /** Timestamp when the proof was created (HLC or unix ms) */
  created_at: string;
  /** Whether this proof has been verified */
  verified: boolean;
}

/**
 * Defines audit requirements for envelope processing.
 */
export interface AuditPolicy {
  /** Unique identifier for this policy */
  policy_id: string;
  /** Whether proof is required for this policy */
  require_proof: boolean;
  /** Level of verification required (e.g., "none", "basic", "strict") */
  verification_level: string;
  /** How many days to retain audit records */
  retention_days: number;
}

/**
 * Represents a complete audit record for an envelope action.
 */
export interface AuditRecord {
  /** Unique identifier for this audit record */
  record_id: string;
  /** The message_id of the envelope being audited */
  envelope_id: string;
  /** The action performed (e.g., "send", "receive", "process") */
  action: string;
  /** ID of the agent/component that performed the action */
  actor_id: string;
  /** When the action occurred (HLC or unix ms) */
  timestamp: string;
  /** Optional proof associated with this action */
  proof?: AuditProof;
}

/**
 * Envelope-like object for audit operations.
 * This is a minimal interface that matches the expected envelope structure.
 */
export interface EnvelopeAuditable {
  message_id?: string;
  producer_id?: string;
  payload?: Uint8Array;
  audit_proof?: Uint8Array;
  [key: string]: any;
}

/**
 * Protocol defining the interface for audit implementations.
 *
 * Implementations can range from no-op (for development) to full
 * ZK-proof based systems (for production).
 */
export interface Auditor {
  /**
   * Create a proof for an envelope action.
   *
   * @param envelope - The envelope to create proof for
   * @param action - The action being performed (e.g., "send", "receive")
   * @returns An AuditProof instance
   */
  createProof(envelope: EnvelopeAuditable, action: string): AuditProof;

  /**
   * Verify the validity of a proof.
   *
   * @param proof - The proof to verify
   * @returns True if proof is valid, false otherwise
   */
  verifyProof(proof: AuditProof): boolean;

  /**
   * Record an audit event.
   *
   * @param envelope - The envelope being audited
   * @param action - The action being performed
   * @param proof - Optional proof to attach to the record
   * @returns The created AuditRecord
   */
  record(envelope: EnvelopeAuditable, action: string, proof?: AuditProof): AuditRecord;

  /**
   * Query audit records for a specific envelope.
   *
   * @param envelope_id - The message_id of the envelope to query
   * @returns List of AuditRecords for this envelope
   */
  query(envelope_id: string): AuditRecord[];
}

/**
 * Generate a UUIDv4 string.
 */
function uuidv4(): string {
  if (typeof (crypto as any).randomUUID === 'function') {
    return (crypto as any).randomUUID();
  }
  // Fallback RFC4122 v4
  const buf = crypto.randomBytes(16);
  buf[6] = (buf[6] & 0x0f) | 0x40;
  buf[8] = (buf[8] & 0x3f) | 0x80;
  const hex = [...buf].map(b => b.toString(16).padStart(2, '0'));
  return (
    hex.slice(0, 4).join('') + '-' +
    hex.slice(4, 6).join('') + '-' +
    hex.slice(6, 8).join('') + '-' +
    hex.slice(8, 10).join('') + '-' +
    hex.slice(10, 16).join('')
  );
}

/**
 * Compute SHA256 hash of an envelope.
 *
 * @param envelope - The envelope to hash
 * @returns Hex-encoded SHA256 hash of the envelope
 */
export function computeEnvelopeHash(envelope: EnvelopeAuditable): string {
  // Create a deterministic representation of the envelope
  const envelopeCopy: any = { ...envelope };

  // Extract binary fields that can't be JSON serialized
  const payload = envelopeCopy.payload;
  const audit_proof = envelopeCopy.audit_proof;
  delete envelopeCopy.payload;
  delete envelopeCopy.audit_proof;

  // Create deterministic JSON representation of non-binary fields
  // Sort keys and use compact JSON to ensure consistency
  const envelopeJson = JSON.stringify(envelopeCopy, Object.keys(envelopeCopy).sort());

  // Create hash
  const hasher = crypto.createHash('sha256');
  hasher.update(envelopeJson, 'utf-8');

  // Include binary fields in hash
  if (payload) {
    if (payload instanceof Uint8Array) {
      hasher.update(payload);
    } else if (typeof payload === 'string') {
      hasher.update(payload, 'utf-8');
    }
  }

  if (audit_proof) {
    if (audit_proof instanceof Uint8Array) {
      hasher.update(audit_proof);
    } else if (typeof audit_proof === 'string') {
      hasher.update(audit_proof, 'utf-8');
    }
  }

  return hasher.digest('hex');
}

/**
 * Create a simple hash-based proof for an envelope.
 *
 * This is a basic proof mechanism suitable for testing and development.
 * Production systems should use more sophisticated proof mechanisms
 * (e.g., ZK-proofs, digital signatures).
 *
 * @param envelope - The envelope to create proof for
 * @param actor_id - ID of the actor creating the proof
 * @returns An AuditProof with simple hash-based proof
 */
export function createSimpleProof(envelope: EnvelopeAuditable, actor_id: string): AuditProof {
  // Compute envelope hash
  const envelopeHash = computeEnvelopeHash(envelope);

  // Create proof data: hash of (envelope_hash + actor_id + timestamp)
  const timestamp = nowHlcStub();
  const proofInput = `${envelopeHash}:${actor_id}:${timestamp}`;
  const proofHash = crypto.createHash('sha256').update(proofInput, 'utf-8').digest();

  return {
    proof_id: uuidv4(),
    proof_type: 'simple_hash',
    proof_data: new Uint8Array(proofHash),
    created_at: timestamp,
    verified: false,
  };
}

/**
 * Verify an audit proof against an envelope.
 *
 * @param envelope - The envelope to verify against
 * @param proof - The proof to verify
 * @returns True if proof is valid, false otherwise
 */
export function verifyAuditProof(envelope: EnvelopeAuditable, proof: AuditProof): boolean {
  // Handle different proof types
  if (proof.proof_type === 'noop') {
    return true;
  }

  if (proof.proof_type === 'simple_hash') {
    // For simple hash proofs, we verify that:
    // 1. The proof data is non-empty
    // 2. The proof was created for this envelope (basic sanity check)
    if (proof.proof_data.length === 0) {
      return false;
    }

    // Compute expected hash for verification
    const envelopeHash = computeEnvelopeHash(envelope);

    // In a real implementation, we would reconstruct the proof
    // and compare it with the provided proof_data
    // For now, we do a basic check that the hash is consistent
    // This is a simplified verification - production systems need more rigor

    // Verify the proof data is 32 bytes (SHA256 output)
    if (proof.proof_data.length !== 32) {
      return false;
    }

    return true;
  }

  // Unknown proof type
  return false;
}

/**
 * No-op auditor that does nothing.
 *
 * Useful for development and when audit trail is not required.
 */
export class NoOpAuditor implements Auditor {
  /**
   * Create a minimal no-op proof.
   *
   * @param envelope - The envelope (ignored)
   * @param action - The action (ignored)
   * @returns A minimal no-op proof
   */
  createProof(envelope: EnvelopeAuditable, action: string): AuditProof {
    return {
      proof_id: uuidv4(),
      proof_type: 'noop',
      proof_data: new Uint8Array(0),
      created_at: nowHlcStub(),
      verified: true,
    };
  }

  /**
   * Always returns true for no-op proofs.
   *
   * @param proof - The proof (ignored)
   * @returns Always true
   */
  verifyProof(proof: AuditProof): boolean {
    return true;
  }

  /**
   * Create a minimal audit record without storing it.
   *
   * @param envelope - The envelope being audited
   * @param action - The action being performed
   * @param proof - Optional proof to attach
   * @returns A minimal audit record
   */
  record(envelope: EnvelopeAuditable, action: string, proof?: AuditProof): AuditRecord {
    return {
      record_id: uuidv4(),
      envelope_id: envelope.message_id ?? 'unknown',
      action,
      actor_id: envelope.producer_id ?? 'unknown',
      timestamp: nowHlcStub(),
      proof,
    };
  }

  /**
   * Always returns empty array.
   *
   * @param envelope_id - The envelope ID (ignored)
   * @returns Empty array
   */
  query(envelope_id: string): AuditRecord[] {
    return [];
  }
}

/**
 * In-memory auditor for testing and development.
 *
 * Stores audit records in memory and supports basic querying.
 * Not suitable for production use.
 */
export class InMemoryAuditor implements Auditor {
  private _records: Map<string, AuditRecord[]>;

  /**
   * Initialize with empty record storage.
   */
  constructor() {
    this._records = new Map();
  }

  /**
   * Create a simple hash-based proof.
   *
   * @param envelope - The envelope to create proof for
   * @param action - The action being performed
   * @returns A simple hash-based proof
   */
  createProof(envelope: EnvelopeAuditable, action: string): AuditProof {
    return createSimpleProof(envelope, envelope.producer_id ?? 'unknown');
  }

  /**
   * Verify a proof using the verification module.
   *
   * @param proof - The proof to verify
   * @returns True if proof is valid, false otherwise
   */
  verifyProof(proof: AuditProof): boolean {
    // Handle different proof types
    if (proof.proof_type === 'noop') {
      return true;
    }

    // For simple hash proofs, check that the proof_data is non-empty
    if (proof.proof_type === 'simple_hash') {
      return proof.proof_data.length > 0;
    }

    // Unknown proof types are not verified
    return false;
  }

  /**
   * Create and store an audit record.
   *
   * @param envelope - The envelope being audited
   * @param action - The action being performed
   * @param proof - Optional proof to attach
   * @returns The created audit record
   */
  record(envelope: EnvelopeAuditable, action: string, proof?: AuditProof): AuditRecord {
    const record: AuditRecord = {
      record_id: uuidv4(),
      envelope_id: envelope.message_id ?? 'unknown',
      action,
      actor_id: envelope.producer_id ?? 'unknown',
      timestamp: nowHlcStub(),
      proof,
    };

    // Store the record
    const envelope_id = record.envelope_id;
    if (!this._records.has(envelope_id)) {
      this._records.set(envelope_id, []);
    }
    this._records.get(envelope_id)!.push(record);

    return record;
  }

  /**
   * Query all audit records for a specific envelope.
   *
   * @param envelope_id - The message_id of the envelope to query
   * @returns List of audit records for this envelope
   */
  query(envelope_id: string): AuditRecord[] {
    return this._records.get(envelope_id) ?? [];
  }

  /**
   * Get all audit records (useful for testing).
   *
   * @returns All audit records across all envelopes
   */
  getAllRecords(): AuditRecord[] {
    const allRecords: AuditRecord[] = [];
    for (const records of this._records.values()) {
      allRecords.push(...records);
    }
    return allRecords;
  }

  /**
   * Clear all stored records (useful for testing).
   */
  clear(): void {
    this._records.clear();
  }
}
